shards/egui/src/visual/mod.rs

```diff
@@ -0,0 +1,2521 @@
+              sequence.statements.push(Statement::Pipeline(Pipeline {
+                blocks: vec![Block {
+                  content: BlockContent::Shard(Function {
+                    name: Identifier {
+                      name: "Pass".into(),
+                      namespaces: Vec::new(),
+                    },
+                    params: None,
+                    custom_state: None,
+                  }),
+                  line_info: None,
+                  custom_state: None,
+                }],
+              }));
+
+              self.context.has_changed = true;
+            }
+            ui.button(emoji("💡")).on_hover_text("Ask AI.")
+          })
+          .inner,
+      )
+    } else {
+      self.context.seqs_stack.pop();
+      Some(
+        self
+          .ui
+          .allocate_response(egui::Vec2::ZERO, egui::Sense::hover()),
+      )
+    }
+  }
+
+  fn visit_statement(&mut self, statement: &mut Statement) -> Option<Response> {
+    self
+      .ui
+      .with_layout(egui::Layout::left_to_right(egui::Align::TOP), |ui| {
+        let mut mutator = VisualAst::with_parent_selected(self.context, ui, self.parent_selected);
+        let is_pipeline = match statement {
+          Statement::Assignment(assignment) => {
+            assignment.accept_mut(&mut mutator);
+            false
+          }
+          Statement::Pipeline(pipeline) => {
+            pipeline.accept_mut(&mut mutator);
+            true
+          }
+        };
+        if self.parent_selected && is_pipeline {
+          Some(
+            ui.horizontal(|ui| {
+              if ui
+                .button(emoji("➕"))
+                .on_hover_text("Add new statement.")
+                .clicked()
+              {
+                // add a new statement
+                let pipeline = match statement {
+                  Statement::Pipeline(pipeline) => pipeline,
+                  _ => unreachable!(),
+                };
+                pipeline.blocks.push(Block {
+                  content: BlockContent::Shard(Function {
+                    name: Identifier {
+                      name: "Pass".into(),
+                      namespaces: Vec::new(),
+                    },
+                    params: None,
+                    custom_state: None,
+                  }),
+                  line_info: None,
+                  custom_state: None,
+                });
+
+                self.context.has_changed = true;
+              }
+              ui.button(emoji("💡")).on_hover_text("Ask AI.")
+            })
+            .inner,
+          )
+        } else {
+          Some(ui.allocate_response(egui::Vec2::ZERO, egui::Sense::hover()))
+        }
+      })
+      .inner
+  }
+
+  fn visit_assignment(&mut self, assignment: &mut Assignment) -> Option<Response> {
+    let mut combined_response = None;
+    let (r_a, r_b) = match assignment {
+      Assignment::AssignRef(pipeline, identifier) => (pipeline.accept_mut(self), {
```

Comment:
I'd still go with the same symbols as text, but maybe we can make some nice icons representing `>/=/>>/>=`

Reply:
We could, not sure how much people will associate and use both modes though tbh, but ok let's start with same